builder/builder.go

```diff
@@ -412,97 +426,448 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
 	return errors.Wrapf(os.Remove(tmptarfile), "Failed to remove temp bundle archive: %s", tmptarfile)
 }
 
-// AddBundles will copy the specified upstream bundles from the configured
-// upstream version to the mix-bundles directory. In error event, the int value
-// returned will contain the number of bundles added thus far.
-// bundles: array slice of bundle names
-// force: override bundle in mix-dir when present
-// all: include all CLR bundles. Overrides bundles.
-// git: automatically git commit with bundles added
-func (b *Builder) AddBundles(bundles []string, force bool, allbundles bool, git bool) (int, error) {
-	var bundleAddCount int
+// getBundlePath returns the path to the bundle definition file for a given
+// bundle name, or error if it cannot be found. Looks first in local-bundles,
+// then upstream-bundles.
+func (b *Builder) getBundlePath(bundle string) (string, error) {
+	// Check local-bundles
+	path := filepath.Join(b.LocalBundleDir, bundle)
+	if _, err := os.Stat(path); err == nil {
+		return path, nil
+	}
+
+	// Check upstream-bundles
+	path = filepath.Join(getUpstreamBundlesPath(b.UpstreamVer), bundle)
+	if _, err := os.Stat(path); err == nil {
+		return path, nil
+	}
+
+	return "", errors.Errorf("Cannot find bundle %s in local or upstream bundles", bundle)
+}
+
+// isLocalBundle checks a bundle filepath is local
+func (b *Builder) isLocalBundle(path string) bool {
+	return strings.HasPrefix(path, b.LocalBundleDir)
+}
+
+type stringSet map[string]bool
+
+func getStringSetKeys(set stringSet) []string {
+	keys := make([]string, len(set))
+	i := 0
+	for k := range set {
+		keys[i] = k
+		i++
+	}
+	return keys
+}
+
+func getStringSetKeysSorted(set stringSet) []string {
+	keys := getStringSetKeys(set)
+	sort.Strings(keys)
+	return keys
+}
+
+func getBundleSetKeysSorted(set bundleSet) []string {
+	keys := make([]string, len(set))
+	i := 0
+	for k := range set {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+// getMixBundlesListAsSet reads in the Mix Bundles List file and returns the
+// resultant set of unique bundle names. If the "mixbundles" file does not
+// exist or is empty, an empty set is returned.
+func (b *Builder) getMixBundlesListAsSet() (stringSet, error) {
+	set := make(stringSet)
+
+	bundles, err := helpers.ReadFileAndSplit("mixbundles")
+	if os.IsNotExist(err) {
+		return set, nil
+	} else if err != nil {
+		return nil, errors.Wrap(err, "Failed to read in Mix Bundle List")
+	}
+
+	for _, bundle := range bundles {
+		set[bundle] = true
```

Comment:
>Ignore anything after #.

See my comment below about why I avoided supporting comments.

>Ignore whitespace only lines.

That's true, I need to skip these either way

>Use strings.TrimSpace on the remaining lines.

Also a good catch.

>Optional: complain about invalid bundle names (we have a regex elsewhere) and tell the filename:linenumber in the error.

These now get immediately parsed as full bundles (part of the force push you missed), so that happens now.